File: apps/admin_backend/src/modules/identity/entities/audit-log.entity.ts
import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { Tenant } from '../../tenant/entities/tenant.entity';
import { User } from './user.entity';

@Entity('audit_logs')
@Index(
  'uq_audit_stream_sequence_active',
  ['tenant_id', 'device_id', 'user_id', 'sequence_no'],
  {
    unique: true,
    where: `forensic_status = 'ACTIVE'`,
  },
)
export class AuditLog {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  tenant_id: string;

  @ManyToOne(() => Tenant)
  @JoinColumn({ name: 'tenant_id' })
  tenant: Tenant;

  @Column()
  user_id: string;

  @ManyToOne(() => User)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @Column()
  action: string;

  @Column({ nullable: true })
  target_type: string;

  @Column({ nullable: true })
  target_id: string;

  @Column()
  device_id: string;

  @Column()
  sequence_no: number;

  @Column()
  prev_hash: string;

  @Column()
  entry_hash: string;

  @Column({ nullable: true })
  metodo_autorizacion: string;

  @Column({ nullable: true })
  usuario_autorizador_id: string;

  @CreateDateColumn({ type: 'timestamptz' })
  timestamp: Date;

  @Column({ type: 'jsonb', default: {} })
  metadata: Record<string, unknown>;

  @Column({ default: 'ACTIVE' })
  forensic_status: string;
}

File: apps/pos_app/lib/data/models/audit_log_entity.dart
import 'package:floor/floor.dart';

@Entity(tableName: 'audit_logs')
class AuditLogEntity {
  @PrimaryKey(autoGenerate: true)
  final int? id;
  @ColumnInfo(name: 'user_id')
  final String userId;
  final String action;
  final String timestamp; // ISO8601
  @ColumnInfo(name: 'device_id')
  final String deviceId;
  final String? metadata;
  @ColumnInfo(name: 'is_synced')
  final bool isSynced;
  @ColumnInfo(name: 'sequence_no')
  final int sequenceNo;
  @ColumnInfo(name: 'prev_hash')
  final String prevHash;
  @ColumnInfo(name: 'entry_hash')
  final String entryHash;
  @ColumnInfo(name: 'metodo_autorizacion')
  final String? metodoAutorizacion;
  @ColumnInfo(name: 'usuario_autorizador_id')
  final String? usuarioAutorizadorId;
  @ColumnInfo(name: 'remote_ref_uuid')
  final String remoteRefUuid;
  @ColumnInfo(name: 'hash_version')
  final String? hashVersion;
  @ColumnInfo(name: 'has_metodo_autorizacion')
  final bool? hasMetodoAutorizacion;
  @ColumnInfo(name: 'has_usuario_autorizador_id')
  final bool? hasUsuarioAutorizadorId;

  AuditLogEntity({
    this.id,
    required this.userId,
    required this.action,
    required this.timestamp,
    required this.deviceId,
    this.metadata,
    this.isSynced = false,
    required this.sequenceNo,
    required this.prevHash,
    required this.entryHash,
    this.metodoAutorizacion,
    this.usuarioAutorizadorId,
    required this.remoteRefUuid,
    this.hashVersion,
    this.hasMetodoAutorizacion,
    this.hasUsuarioAutorizadorId,
  });
}
